File: src/types/equipment/item/AnimalCare.ts
import {
  Entity,
  Enum,
  EnumCase,
  IncludeIdentifier,
  Object,
  Required,
  TypeAlias,
} from "tsondb/schema/dsl"
import { src } from "../../source/_PublicationRef.js"
import { Cost, DefaultItemTranslations, FixedCost, Weight } from "./_Item.js"

export const AnimalCare = Entity(import.meta.url, {
  name: "AnimalCare",
  namePlural: "AnimalCare",
  type: () =>
    Object({
      type: Required({
        comment: "Values depending on whether the animal care is feed.",
        type: IncludeIdentifier(AnimalCareType),
      }),
      src,
      translations: DefaultItemTranslations("AnimalCare"),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const AnimalCareType = Enum(import.meta.url, {
  name: "AnimalCareType",
  comment: "Values depending on whether the animal care is feed.",
  values: () => ({
    General: EnumCase({ type: IncludeIdentifier(GeneralAnimalCare) }),
    Feed: EnumCase({ type: IncludeIdentifier(AnimalFeed) }),
  }),
})

const GeneralAnimalCare = TypeAlias(import.meta.url, {
  name: "GeneralAnimalCare",
  type: () =>
    Object({
      cost: Required({
        comment: "The cost in silverthalers.",
        type: IncludeIdentifier(Cost),
      }),
      weight: Required({
        comment: "The weight in kg.",
        type: IncludeIdentifier(Weight),
      }),
    }),
})

const AnimalFeed = TypeAlias(import.meta.url, {
  name: "AnimalFeed",
  type: () =>
    Object({
      cost: Required({
        comment: "The cost in silverthalers.",
        type: IncludeIdentifier(AnimalFeedCost),
      }),
    }),
})

const AnimalFeedCost = Enum(import.meta.url, {
  name: "AnimalFeedCost",
  values: () => ({
    PerWeek: EnumCase({ type: IncludeIdentifier(FixedCost) }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/equipment/item/AnimalCare.swift
import FileDB

@Model
public struct AnimalCare {
  name: "AnimalCare",
  namePlural: "AnimalCare",
  type: () =>
    Object({

  /// Values depending on whether the animal care is feed.
  @Relationship(AnimalCareType)
  let type: AnimalCareType.ID

    /// The publications where you can find the entry.
    let src: PublicationRefs
      translations: DefaultItemTranslations("AnimalCare"),
    }),
}

/// Values depending on whether the animal care is feed.
@ModelEnum
public enum AnimalCareType {
    case General(IncludeIdentifier(GeneralAnimalCare))
    case Feed(IncludeIdentifier(AnimalFeed))
}

@Embedded
public struct GeneralAnimalCare {

  /// The cost in silverthalers.
  @Relationship(Cost)
  let cost: Cost.ID

  /// The weight in kg.
  @Relationship(Weight)
  let weight: Weight.ID
  }

@Embedded
public struct AnimalFeed {

  /// The cost in silverthalers.
  @Relationship(AnimalFeedCost)
  let cost: AnimalFeedCost.ID
  }

@ModelEnum
public enum AnimalFeedCost {
    case PerWeek(IncludeIdentifier(FixedCost))
}
